src/main/java/org/jabref/gui/maintable/MainTableDataModel.java

```diff
@@ -102,6 +104,35 @@ public MainTableDataModel(BibDatabaseContext context,
         resultSizeProperty.bind(Bindings.size(entriesFiltered.filtered(entry -> entry.matchCategory().isEqualTo(MatchCategory.MATCHING_SEARCH_AND_GROUPS).get())));
         // We need to wrap the list since otherwise sorting in the table does not work
         entriesFilteredAndSorted = new SortedList<>(entriesFiltered);
+
+        initializeEntryIndexMap();
+    }
+
+    private void initializeEntryIndexMap() {
+        for (int i = 0; i < allEntries.size(); i++) {
+            entryIndexMap.put(allEntries.get(i).getId(), i);
+        }
+
+        allEntries.addListener((ListChangeListener.Change<? extends BibEntry> c) -> {
+            while (c.next()) {
+                if (c.wasPermutated()) {
+                    for (int i = c.getFrom(); i < c.getTo(); ++i) {
+                        entryIndexMap.put(allEntries.get(i).getId(), i);
+                    }
+                } else {
+                    if (c.wasRemoved()) {
+                        for (BibEntry removedEntry : c.getRemoved()) {
+                            entryIndexMap.remove(removedEntry.getId());
```

Comment:
I think, all indexes AFTER that number will decrease by one?

Reply:
Reverted, and used binary search to find the index [`94b27eb` (#11542)](https://github.com/JabRef/jabref/pull/11542/commits/94b27eb9350d821b57022e3348834b3aed11e9b2).